<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Primer design • primex</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">primex</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Usage.html">Primer design</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dieterich-lab/primex">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Primer design</h1>
                        <h4 class="author">Alexey Uvarovskii</h4>
            
            <h4 class="date">2018-01-17</h4>
          </div>

    
    
<div class="contents">
<style>
body {
  max-width: 900px;
}
</style>
<div id="primer-design-for-transcripts-and-arbitrary-sequences" class="section level2">
<h2 class="hasAnchor">
<a href="#primer-design-for-transcripts-and-arbitrary-sequences" class="anchor"></a>Primer design for transcripts and arbitrary sequences</h2>
<p>The package allows to run the Primer3 with user specified parameters. There is a shortcut for design, in which one of the primers must overlap with the splice junction.</p>
<div id="a-quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#a-quick-start" class="anchor"></a>A quick start</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(primex)
exon1 &lt;-<span class="st"> </span><span class="kw">paste0</span>(
  <span class="st">"CTCACCATGGATGATGATATCGCCGCGCTCGTCGTCGACAACGGCTCCGGCATGTGCAAG"</span>,
  <span class="st">"GCCGGCTTCGCGGGCGACGATGCCCCCCGGGCCGTCTTCCCCTCCATCGTGGGGCGCCCC"</span>,
  <span class="st">"AGGCACCAG"</span>)
exon2 &lt;-<span class="st"> </span><span class="kw">paste0</span>(
 <span class="st">"GGCGTGATGGTGGGCATGGGTCAGAAGGATTCCTATGTGGGCGACGAGGCCCAGAGCAAG"</span>,
 <span class="st">"AGAGGCATCCTCACCCTGAAGTACCCCATCGAGCACGGCATCGTCACCAACTGGGACGAC"</span>,
 <span class="st">"ATGGAGAAAATCTGGCACCACACCTTCTACAATGAGCTGCGTGTGGCTCCCGAGGAGCAC"</span>,
 <span class="st">"CCCGTGCTGCTGACCGAGGCCCCCCTGAACCCCAAGGCCAACCGCGAGAAGATGACCCAG"</span>) 

seqOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seqSettings.html">seqSettings</a></span>(
  <span class="dt">seqId =</span> <span class="st">"example1"</span>,
  <span class="dt">seq =</span> <span class="kw">c</span>(exon1, exon2))

primers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/design.html">design</a></span>(seqOpts)</code></pre></div>
</div>
<div id="a-quick-start-with-annotations" class="section level3">
<h3 class="hasAnchor">
<a href="#a-quick-start-with-annotations" class="anchor"></a>A quick start with annotations</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(primex)  

<span class="kw">library</span>(EnsDb.Hsapiens.v86)
<span class="kw">library</span>(BSgenome.Hsapiens.NCBI.GRCh38)
db &lt;-<span class="st"> </span>EnsDb.Hsapiens.v86
bsg &lt;-<span class="st"> </span>BSgenome.Hsapiens.NCBI.GRCh38

<span class="co"># define sequences</span>
geneName &lt;-<span class="st"> "BCL6"</span>
exonsByTx &lt;-<span class="st"> </span>ensembldb::<span class="kw">exonsBy</span>(db, <span class="dt">filter =</span> <span class="kw">GenenameFilter</span>(geneName))
<span class="co"># find unique splice junctions </span>
sjExonPairs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/selectPairs.html">selectPairs</a></span>(exonsByTx)
<span class="co"># or report all the pairs</span>
<span class="co"># sjExonPairs &lt;- selectPairs(exonsByTx, tolerance = -1)</span>

<span class="co"># take your transcript of interest</span>
exseqs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addSeq.html">addSeq</a></span>(sjExonPairs$ENST00000419510, bsg)

<span class="co"># run Primer3 for one of splice junction pair</span>
seqOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seqSettings.html">seqSettings</a></span>(<span class="dt">gr =</span> exseqs$<span class="st">`</span><span class="dt">ENSE00001666929|ENSE00001377234</span><span class="st">`</span>)
primers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/design.html">design</a></span>(seqOpts)</code></pre></div>
</div>
</div>
<div id="step-by-step" class="section level2">
<h2 class="hasAnchor">
<a href="#step-by-step" class="anchor"></a>Step by step</h2>
<div id="find-unique-exons" class="section level3">
<h3 class="hasAnchor">
<a href="#find-unique-exons" class="anchor"></a>Find unique exons</h3>
<p>Let us consider human BCL6 gene, which has a lot of isoforms. To identify which exons to use for the primer design, we need a <code>GRanges</code> list with the gene model, such as:</p>
<ol style="list-style-type: decimal">
<li>Names of the list items correspond to the transcript names.</li>
<li>Exons are ordered according to their rank (<code>exon_rank</code> in the <code>ensembldb</code> packages).</li>
<li>A metadata column <code>exon_id</code> must be present in the GRanges objects.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">primex::bcl6exons[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## GRanges object with 10 ranges and 3 metadata columns:
##        seqnames                 ranges strand |         exon_id
##           &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##    [1]        3 [187736090, 187736569]      - | ENSE00001512051
##    [2]        3 [187734869, 187734907]      - | ENSE00001377234
##    [3]        3 [187733533, 187733703]      - | ENSE00003458337
##    [4]        3 [187731709, 187731930]      - | ENSE00000781579
##    [5]        3 [187729050, 187730021]      - | ENSE00003602776
##    [6]        3 [187728360, 187728544]      - | ENSE00003532311
##    [7]        3 [187726731, 187726898]      - | ENSE00003627638
##    [8]        3 [187725499, 187725629]      - | ENSE00003540935
##    [9]        3 [187724941, 187725078]      - | ENSE00003468191
##   [10]        3 [187722398, 187722601]      - | ENSE00001928038
##          gene_name exon_rank
##        &lt;character&gt; &lt;integer&gt;
##    [1]        BCL6         1
##    [2]        BCL6         2
##    [3]        BCL6         3
##    [4]        BCL6         4
##    [5]        BCL6         5
##    [6]        BCL6         6
##    [7]        BCL6         7
##    [8]        BCL6         8
##    [9]        BCL6         9
##   [10]        BCL6        10
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>In order to distinguish isoforms using PCR, we look for unique splice junctions, i.e. unique exon pairs. <strong>This is the default behaviour. </strong> For every transcript we may receive several such candidates.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sjExonPairs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/selectPairs.html">selectPairs</a></span>(primex::bcl6exons)</code></pre></div>
<p>Since we included transcripts with all types of support, there are not so many unique junctions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(sjExonPairs, length)</code></pre></div>
<pre><code>## ENST00000232014 ENST00000406870 ENST00000419510 ENST00000430339 
##               0               0               0               0 
## ENST00000438077 ENST00000450123 ENST00000470319 ENST00000479110 
##               1               0               0               0 
## ENST00000480458 ENST00000496823 ENST00000621333 
##               0               1               0</code></pre>
<p>The gene model can be visualised using <code>primex</code> as well:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotSegments.html">plotSegments</a></span>(primex::bcl6exons)</code></pre></div>
<p><img src="Usage_files/figure-html/unnamed-chunk-7-1.png" width="70%"></p>
</div>
<div id="provide-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#provide-sequences" class="anchor"></a>Provide sequences</h3>
<p>If there is available BSgenome package, one may retrieve exonic sequences from it. We implemented a functionality to ease this interface:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressPackageStartupMessages</span>(
  <span class="kw">library</span>(BSgenome.Hsapiens.NCBI.GRCh38)
)
bsg &lt;-<span class="st"> </span>BSgenome.Hsapiens.NCBI.GRCh38

exonPairs &lt;-<span class="st"> </span>sjExonPairs$ENST00000438077
exseqs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addSeq.html">addSeq</a></span>(exonPairs, bsg)
exseqs$<span class="st">`</span><span class="dt">ENSE00001752117|ENSE00001377234</span><span class="st">`</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 4 metadata columns:
##       seqnames                 ranges strand |         exon_id   gene_name
##          &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;
##   [1]        3 [187737597, 187737944]      - | ENSE00001752117        BCL6
##   [2]        3 [187734869, 187734907]      - | ENSE00001377234        BCL6
##       exon_rank
##       &lt;integer&gt;
##   [1]         1
##   [2]         2
##                                                                                                                                                                                                                                                                                                                                                                seq
##                                                                                                                                                                                                                                                                                                                                                        &lt;character&gt;
##   [1] TCCTGGCGGCTGCCGGGGCCGGCAGCTTCCTGGAAAGTTACTTCTTGTTGGAGCCGGCAATAGGCAAAGTGTGTTCTGTGAAATCCGCAGAGCCGAGATTGACTACGTTCCGGGAATGAGAGGGCTGTGTCATTCCCCTCATTGCCTGCTGCATTGACGGCGTAACAGGAAAAAAAAAAAAAAAAAAAAAAAGGCATACACACAATGTTAGCTACCGAGGTGCACCAAAAGTTTTATTTGAATGTAACTAAATTAAGGGCCGCCACGGTCATACCACAAAGCCTCCGTCGCGCCTTGCGGGGGGCTTCGAGGTGGATCGCCCAGGGGCGGGCAGTCCCTGGAAGACAG
##   [2]                                                                                                                                                                                                                                                                                                                      GTTTTGAGCAAAATTTTGGACTGTGAAGCAAGGCATTGG
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
</div>
<div id="run-primer3" class="section level3">
<h3 class="hasAnchor">
<a href="#run-primer3" class="anchor"></a>Run Primer3</h3>
<p>The <em>Primer3</em> program has a lot of parameters. There are two sets, <em>SEQUENCE</em> parameters and <em>PRIMER</em> parameters. The <em>SEQUENCE</em> parameters describe such items, as a sequence identificator, the sequence of nucleotides itself, splice junction position, etc. One may start from the <code>GRanges</code> object, which has a column <code>seq</code> with a nucleotide sequence for every item:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seqOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seqSettings.html">seqSettings</a></span>(<span class="dt">gr =</span> exseqs$<span class="st">`</span><span class="dt">ENSE00001752117|ENSE00001377234</span><span class="st">`</span>)
res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/design.html">design</a></span>(seqOpts)
<span class="co"># take first two pairs: Primer3 reports by default maximum 5 pairs</span>
knitr::<span class="kw">kable</span>(<span class="kw">t</span>(res$primers)[,<span class="dv">1</span>:<span class="dv">2</span>])</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="left">0</th>
<th align="left">1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>PRIMER_PAIR_PENALTY</td>
<td align="left">0.651478</td>
<td align="left">0.652652</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_PENALTY</td>
<td align="left">0.066262</td>
<td align="left">0.067436</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_PENALTY</td>
<td align="left">0.585216</td>
<td align="left">0.585216</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_SEQUENCE</td>
<td align="left">GCTGTGTCATTCCCCTCATT</td>
<td align="left">GATTGACTACGTTCCGGGAA</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_SEQUENCE</td>
<td align="left">TGCTCAAAACCTGTCTTCCA</td>
<td align="left">TGCTCAAAACCTGTCTTCCA</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT</td>
<td align="left">123,20</td>
<td align="left">96,20</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT</td>
<td align="left">357,20</td>
<td align="left">357,20</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_TM</td>
<td align="left">59.934</td>
<td align="left">59.933</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_TM</td>
<td align="left">59.415</td>
<td align="left">59.415</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_GC_PERCENT</td>
<td align="left">50.000</td>
<td align="left">50.000</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_GC_PERCENT</td>
<td align="left">45.000</td>
<td align="left">45.000</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_SELF_ANY_TH</td>
<td align="left">0.00</td>
<td align="left">0.73</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_SELF_ANY_TH</td>
<td align="left">0.00</td>
<td align="left">0.00</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_SELF_END_TH</td>
<td align="left">0.00</td>
<td align="left">0.73</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_SELF_END_TH</td>
<td align="left">0.00</td>
<td align="left">0.00</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_HAIRPIN_TH</td>
<td align="left">0.00</td>
<td align="left">0.00</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_HAIRPIN_TH</td>
<td align="left">0.00</td>
<td align="left">0.00</td>
</tr>
<tr class="even">
<td>PRIMER_LEFT_END_STABILITY</td>
<td align="left">6.9000</td>
<td align="left">9.7000</td>
</tr>
<tr class="odd">
<td>PRIMER_RIGHT_END_STABILITY</td>
<td align="left">8.5000</td>
<td align="left">8.5000</td>
</tr>
<tr class="even">
<td>PRIMER_PAIR_COMPL_ANY_TH</td>
<td align="left">0.00</td>
<td align="left">0.00</td>
</tr>
<tr class="odd">
<td>PRIMER_PAIR_COMPL_END_TH</td>
<td align="left">0.00</td>
<td align="left">0.00</td>
</tr>
<tr class="even">
<td>PRIMER_PAIR_PRODUCT_SIZE</td>
<td align="left">235</td>
<td align="left">262</td>
</tr>
</tbody>
</table>
</div>
<div id="result-in-granges" class="section level3">
<h3 class="hasAnchor">
<a href="#result-in-granges" class="anchor"></a>Result in GRanges</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grPrimers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/toGRanges.html">toGRanges</a></span>(res$primers, exseqs$<span class="st">`</span><span class="dt">ENSE00001752117|ENSE00001377234</span><span class="st">`</span>)
<span class="kw">names</span>(grPrimers) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"primers"</span>, <span class="dv">1</span>:<span class="dv">2</span>)
<span class="kw"><a href="../reference/plotSegments.html">plotSegments</a></span>(bcl6exons, <span class="kw"><a href="../reference/primersToList.html">primersToList</a></span>(grPrimers))</code></pre></div>
<p><img src="Usage_files/figure-html/unnamed-chunk-10-1.png" width="70%"></p>
</div>
<div id="more-on-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#more-on-parameters" class="anchor"></a>More on parameters</h3>
<p>There are some shortcuts for the most common parameters. They can be combined with the pipe operator <code>%&gt;%</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p3Opts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/p3Settings.html">p3Settings</a></span>() %&gt;%
<span class="st">  </span><span class="kw"><a href="../reference/primerSize.html">primerSize</a></span>(<span class="dt">min =</span> <span class="dv">18</span>, <span class="dt">optimal =</span> <span class="dv">20</span>, <span class="dt">max =</span> <span class="dv">22</span>) %&gt;%
<span class="st">  </span><span class="kw"><a href="../reference/primerTm.html">primerTm</a></span>(<span class="dt">min =</span> <span class="dv">57</span>, <span class="dt">optimal =</span> <span class="dv">61</span>, <span class="dt">max =</span> <span class="dv">64</span>) %&gt;%
<span class="st">  </span><span class="kw"><a href="../reference/productSize.html">productSize</a></span>(<span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">400</span>))</code></pre></div>
<p>or they can be used on their own:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p3Opts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/productSize.html">productSize</a></span>(p3Opts, <span class="kw">c</span>(<span class="dv">200</span>,<span class="dv">300</span>))</code></pre></div>
<p>All the parameters can be explicitly specified:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seqOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seqSettings.html">seqSettings</a></span>(<span class="dt">seqId =</span> <span class="st">"example2"</span>, <span class="dt">seq =</span> <span class="kw">c</span>(<span class="st">"AAATGCTGAAGGT"</span>))
seqOpts$SEQUENCE_TARGET &lt;-<span class="st"> "37,21"</span>
<span class="kw">str</span>(seqOpts)</code></pre></div>
<pre><code>## List of 17
##  $ SEQUENCE_EXCLUDED_REGION           : NULL
##  $ SEQUENCE_INCLUDED_REGION           : NULL
##  $ SEQUENCE_PRIMER_REVCOMP            : NULL
##  $ SEQUENCE_FORCE_LEFT_END            : NULL
##  $ SEQUENCE_INTERNAL_EXCLUDED_REGION  : NULL
##  $ SEQUENCE_QUALITY                   : NULL
##  $ SEQUENCE_FORCE_LEFT_START          : NULL
##  $ SEQUENCE_INTERNAL_OLIGO            : NULL
##  $ SEQUENCE_START_CODON_POSITION      : NULL
##  $ SEQUENCE_FORCE_RIGHT_END           : NULL
##  $ SEQUENCE_OVERLAP_JUNCTION_LIST     : NULL
##  $ SEQUENCE_TARGET                    : chr "37,21"
##  $ SEQUENCE_FORCE_RIGHT_START         : NULL
##  $ SEQUENCE_PRIMER                    : NULL
##  $ SEQUENCE_TEMPLATE                  : chr "AAATGCTGAAGGT"
##  $ SEQUENCE_ID                        : chr "example2"
##  $ SEQUENCE_PRIMER_PAIR_OK_REGION_LIST: NULL</code></pre>
<p>The parameters must be set as character strings.</p>
<p>The run settings are set using <code>p3Settings</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">primerOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/p3Settings.html">p3Settings</a></span>()
primerOpts$PRIMER_PRODUCT_OPT_SIZE &lt;-<span class="st"> "100,300"</span>
primerOpts$PRIMER_OPT_TM &lt;-<span class="st"> "60.0"</span></code></pre></div>
</div>
</div>
<div id="i-want-to-use-own-executable" class="section level2">
<h2 class="hasAnchor">
<a href="#i-want-to-use-own-executable" class="anchor"></a>I want to use own executable</h2>
<p>It is possible to run your own <em>Primer3</em> executable, having provided the needed paths.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">primerOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/p3Settings.html">p3Settings</a></span>(
  <span class="dt">defaultsFile =</span> 
    <span class="st">"/home/tsawyer/primer3/primer3_v1_1_4_default_settings.txt"</span>)
<span class="kw"><a href="../reference/design.html">design</a></span>(
  seqOpts,
  primerOpts,
  <span class="dt">returnStats =</span> <span class="ot">TRUE</span>,
  <span class="dt">path =</span> <span class="kw">list</span>(<span class="dt">primer3 =</span> <span class="st">"/home/tsawyer/primer3/primer3_core"</span>,
              <span class="dt">config  =</span> <span class="st">"/home/tsawyer/primer3/primer3_config"</span>)
)</code></pre></div>
<p>Please, refer to the <a href="http://primer3.sourceforge.net/primer3_manual.htm">Primer3 manual page</a> for more details.</p>
</div>
<div id="no-primers" class="section level2">
<h2 class="hasAnchor">
<a href="#no-primers" class="anchor"></a>No primers?</h2>
<p>Let us have two subsequent exons and we would like to design a pair of primers for them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exon1 &lt;-<span class="st"> </span><span class="kw">paste0</span>(
  <span class="st">"CTCACCATGGATGATGATATCGCCGCGCTCGTCGTCGACAACGGCTCCGGCATGTGCAAG"</span>,
  <span class="st">"GCCGGCTTCGCGGGCGACGATGCCCCCCGGGCCGTCTTCCCCTCCATCGTGGGGCGCCCC"</span>,
  <span class="st">"AGGCACCAG"</span>)
exon2 &lt;-<span class="st"> </span><span class="kw">paste0</span>(
 <span class="st">"GGCGTGATGGTGGGCATGGGTCAGAAGGATTCCTATGTGGGCGACGAGGCCCAGAGCAAG"</span>,
 <span class="st">"AGAGGCATCCTCACCCTGAAGTACCCCATCGAGCACGGCATCGTCACCAACTGGGACGAC"</span>,
 <span class="st">"ATGGAGAAAATCTGGCACCACACCTTCTACAATGAGCTGCGTGTGGCTCCCGAGGAGCAC"</span>,
 <span class="st">"CCCGTGCTGCTGACCGAGGCCCCCCTGAACCCCAAGGCCAACCGCGAGAAGATGACCCAG"</span>) </code></pre></div>
<p>It is straightforward to call Primer3 with the default values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seqOpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seqSettings.html">seqSettings</a></span>(
  <span class="dt">seqId =</span> <span class="st">"example1"</span>,
  <span class="dt">seq =</span> <span class="kw">c</span>(exon1, exon2))

result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/design.html">design</a></span>(seqOpts)
result$primers</code></pre></div>
<pre><code>## NULL</code></pre>
<p>We can see, that there are no primers returned. Let us see, what can be a reason for their rejection:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/diagnose.html">diagnose</a></span>(result)</code></pre></div>
<pre><code>## $PRIMER_LEFT_EXPLAIN
## [1] "considered 2614, GC content failed 159, low tm 124, high tm 2013, long poly-x seq 19, ok 299"
## 
## $PRIMER_RIGHT_EXPLAIN
## [1] "considered 2515, GC content failed 84, low tm 122, high tm 1964, long poly-x seq 32, ok 313"
## 
## $PRIMER_PAIR_EXPLAIN
## [1] "considered 93587, no overlap of required point 93587, ok 0"</code></pre>
<p>By default, <code>design</code> overwrites the option <code>PRIMER_EXPLAIN_FLAG</code>, which allows to see the run statistics.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#primer-design-for-transcripts-and-arbitrary-sequences">Primer design for transcripts and arbitrary sequences</a></li>
      <li><a href="#step-by-step">Step by step</a></li>
      <li><a href="#i-want-to-use-own-executable">I want to use own executable</a></li>
      <li><a href="#no-primers">No primers?</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alexey Uvarovskii.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
